File: tests/typesafety/python/entity_ids_misuse.py
# expect: fails
# title: The swapped bill, rejected by the checker
# misuse-of: entity_ids
from typing import NewType

CustomerId = NewType("CustomerId", str)
ProductId = NewType("ProductId", str)


def bill(customer: CustomerId, product: ProductId) -> str:
    return f"invoice {customer} for one {product}"


def bill_roadster(customer: CustomerId, product: ProductId) -> str:
    return bill(product, customer)  # rejected: both types are wrong

File: tests/typesafety/typescript/entity_ids_misuse.ts
// expect: fails

declare const tenantBrand: unique symbol;
declare const landlordBrand: unique symbol;

type TenantAccount = string & { readonly [tenantBrand]: true };
type LandlordAccount = string & { readonly [landlordBrand]: true };

function transfer(source: TenantAccount, target: LandlordAccount, amountCents: number): string {
  return `move ${amountCents} from ${source} to ${target}`;
}

export function payRent(tenant: TenantAccount, landlord: LandlordAccount): string {
  return transfer(landlord, tenant, 95_000); // rejected: both types are wrong
}
